daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1075,88 +1074,63 @@ private[lf] object SBuiltin {
-        machine: Machine[Q],
-    ): Control[Nothing] = {
-      def coid = getSContractId(args, 0)
-      val (actualTemplateId, record) = getSAnyContract(args, 1)
-      if (actualTemplateId != templateId) {
-        Control.Error(IE.WronglyTypedContract(coid, templateId, actualTemplateId))
-      } else {
-        Control.Value(record)
-      }
-    }
-  }
-
-  // SBPromoteAnyContract(templateId): ContractId actualTemplateId -> Any -> templateId
-  // Fails unless actualTemplateId is a predecessor of templateId.
-  final case class SBPromoteAnyContract(
+  final case class SBCastAnyContract(
       templateId: TypeConName,
-      fields: ImmArray[(Ast.FieldName, Ast.Type)],
-      acceptedTemplateIds: List[TypeConName],
+      targetFieldsAndTypes: ImmArray[(Ast.FieldName, Ast.Type)],
+      // TODO: https://github.com/digital-asset/daml/issues/17082
+      // - allowUpgradesAndDowngrades will come from a feature flag
+      allowUpgradesAndDowngrades: Boolean,
   ) extends SBuiltin(2) {
+
+    // TODO: https://github.com/digital-asset/daml/issues/17082
+    // - we currently make no use of the types
+    val targetFields = targetFieldsAndTypes.map(_._1)
+
     override private[speedy] def execute[Q](
         args: util.ArrayList[SValue],
         machine: Machine[Q],
     ): Control[Nothing] = {
-
-      // TODO https://github.com/digital-asset/daml/issues/16151
-      // debug calls to SBPromoteAnyContract
-      // println(s"**SBPromoteAnyContract: \n- templateId=${templateId}\n- acceptedTemplateIds=${acceptedTemplateIds}");
-
-      // TODO https://github.com/digital-asset/daml/issues/16151
-      // perhaps compute `acceptedTemplateIds` here, dynamically when execute is called
-      // And then be can be more confident that all relavant package have been compiled.
-
-      def coid = getSContractId(args, 0)
       val (actualTemplateId, record) = getSAnyContract(args, 1)
-
-      if ((templateId +: acceptedTemplateIds).contains(actualTemplateId)) {
-        // Here we extend values of predecessor template types by adding the
-        // right number of 'None's for missing 'Optional' fields
-        // TODO: https://github.com/digital-asset/daml/issues/16151
-        // For the PoC, this assumes field order is preserved by later versions
-        // and that new fields are only added at the end.
-
-        // TODO: https://github.com/digital-asset/daml/issues/16151
-        // This `SBPromoteAnyContract' code will be reworked shortly.
-        // add-Node/drop-None will be moved to importValue
-
-        val newFields = fields.iterator.drop(record.values.size)
-        val badNewFields = newFields.filter {
-          case (name @ _, TOptional(_)) => false
-          case _ => true
-        }
-        if (badNewFields.isEmpty) {
-          Control.Value(
+      if (actualTemplateId == templateId) {
+        // No upgrade or downgrade required
+        Control.Value(record)
+      } else {
+        assert(actualTemplateId != templateId)
```

Comment:
Given we are within a conditional, do we need this assertion?

Reply:
No, we don't need these. 
I was just being careful and trying to make the code easier to follow.